File: services/runtime/src/rino_runtime/contracts/validation.py
"""Canonical schema validation for IPC messages, project documents, and the registry.

Validation is format aware so identifiers and timestamps are enforced, and diagnostics
are bounded so no document or payload content reaches logs or error responses.
"""

from __future__ import annotations

from typing import Any, Final, Protocol, cast

from jsonschema import Draft202012Validator, FormatChecker
from jsonschema.exceptions import ValidationError

from rino_runtime.contracts.generated.rino_graph_v1_schema import RINO_GRAPH_V1_SCHEMA
from rino_runtime.contracts.generated.rino_ipc_v1_schema import RINO_IPC_V1_SCHEMA
from rino_runtime.contracts.generated.rino_registry_v1_schema import (
    RINO_REGISTRY_V1_SCHEMA,
)

MAXIMUM_DIAGNOSTIC_LENGTH: Final[int] = 512
_MAXIMUM_REPORTED_ERRORS: Final[int] = 4


class _SchemaValidator(Protocol):
    def is_valid(self, instance: object) -> bool: ...

    def iter_errors(self, instance: object) -> object: ...


def _build_validator(schema: dict[str, Any]) -> _SchemaValidator:
    Draft202012Validator.check_schema(schema)
    return cast(
        "_SchemaValidator",
        Draft202012Validator(schema, format_checker=FormatChecker()),
    )


_MESSAGE_VALIDATOR: Final[_SchemaValidator] = _build_validator(RINO_IPC_V1_SCHEMA)
_DOCUMENT_VALIDATOR: Final[_SchemaValidator] = _build_validator(RINO_GRAPH_V1_SCHEMA)
_REGISTRY_VALIDATOR: Final[_SchemaValidator] = _build_validator(RINO_REGISTRY_V1_SCHEMA)
_DEFINITION_VALIDATORS: Final[dict[tuple[str, str], _SchemaValidator]] = {}


def _definition_validator(
    contract: str, schema: dict[str, Any], definition: str
) -> _SchemaValidator:
    cached = _DEFINITION_VALIDATORS.get((contract, definition))
    if cached is not None:
        return cached
    if definition not in schema["$defs"]:
        raise KeyError(f"Unknown canonical definition: {contract}.{definition}.")
    definition_schema: dict[str, Any] = {
        "$schema": schema["$schema"],
        "$ref": f"#/$defs/{definition}",
        "$defs": schema["$defs"],
    }
    validator = _build_validator(definition_schema)
    _DEFINITION_VALIDATORS[(contract, definition)] = validator
    return validator


def _describe(validator: _SchemaValidator, value: object) -> str:
    errors = cast(
        "list[ValidationError]",
        list(validator.iter_errors(value)),  # type: ignore[arg-type]
    )
    if not errors:
        return ""
    described = "; ".join(
        f"{_instance_path(error)} {error.validator}"
        for error in errors[:_MAXIMUM_REPORTED_ERRORS]
    )
    return described[:MAXIMUM_DIAGNOSTIC_LENGTH]


def is_valid_message(value: object) -> bool:
    """Reports whether a decoded JSON value satisfies the canonical envelope schema."""
    return _MESSAGE_VALIDATOR.is_valid(value)


def is_valid_payload(definition: str, value: object) -> bool:
    """Reports whether a payload or result satisfies one named canonical definition."""
    return _definition_validator("ipc", RINO_IPC_V1_SCHEMA, definition).is_valid(value)


def describe_message_errors(value: object) -> str:
    """Returns a bounded structural diagnostic that never echoes payload content."""
    return _describe(_MESSAGE_VALIDATOR, value)


def is_valid_project_document(value: object) -> bool:
    """Reports whether a value satisfies the canonical project document schema."""
    return _DOCUMENT_VALIDATOR.is_valid(value)


def describe_project_document_errors(value: object) -> str:
    """Returns a bounded structural diagnostic that never echoes document content."""
    return _describe(_DOCUMENT_VALIDATOR, value)


def is_valid_graph_definition(definition: str, value: object) -> bool:
    """Reports whether a value satisfies one named definition of the graph schema."""
    return _definition_validator("graph", RINO_GRAPH_V1_SCHEMA, definition).is_valid(
        value
    )


def is_valid_registry_snapshot(value: object) -> bool:
    """Reports whether a value satisfies the canonical registry snapshot schema."""
    return _REGISTRY_VALIDATOR.is_valid(value)


def describe_registry_snapshot_errors(value: object) -> str:
    """Returns a bounded structural diagnostic that never echoes registry content."""
    return _describe(_REGISTRY_VALIDATOR, value)


def is_valid_registry_definition(definition: str, value: object) -> bool:
    """Reports whether a value satisfies one named definition of the registry schema."""
    return _definition_validator(
        "registry", RINO_REGISTRY_V1_SCHEMA, definition
    ).is_valid(value)


def _instance_path(error: ValidationError) -> str:
    if not error.absolute_path:
        return "/"
    return "/" + "/".join(str(part) for part in error.absolute_path)

File: packages/contracts-ts/src/validation.ts
import {
  definitionValidators as diagnosticsDefinitionValidators,
  rootValidator as diagnosticReportValidator,
} from "./generated/rino-diagnostics-v1.validators.js";
import type { RinoGraphDiagnosticReportV1 } from "./generated/rino-diagnostics-v1.types";
import {
  definitionValidators as graphDefinitionValidators,
  rootValidator as projectDocumentValidator,
} from "./generated/rino-graph-v1.validators.js";
import type {
  GraphDocumentV1,
  ProjectManifestV1,
  RinoProjectDocumentV1,
} from "./generated/rino-graph-v1.types";
import {
  definitionValidators as ipcDefinitionValidators,
  rootValidator as messageValidator,
} from "./generated/rino-ipc-v1.validators.js";
import type { RinoIpcMessageV1 } from "./generated/rino-ipc-v1.types";
import {
  definitionValidators as registryDefinitionValidators,
  rootValidator as registrySnapshotValidator,
} from "./generated/rino-registry-v1.validators.js";
import type { RinoNodeRegistrySnapshotV1 } from "./generated/rino-registry-v1.types";
import type { PayloadDefinitionName } from "./message-families";

const MAXIMUM_REPORTED_ERRORS = 4;

interface ValidationError {
  readonly instancePath: string;
  readonly keyword: string;
}

interface StaticValidator {
  (value: unknown): boolean;
  errors?: readonly ValidationError[] | null;
}

const schemaDefinitionValidators = {
  diagnostics: diagnosticsDefinitionValidators,
  graph: graphDefinitionValidators,
  ipc: ipcDefinitionValidators,
  registry: registryDefinitionValidators,
} as const;

function definitionValidator(
  definitions: ReadonlyMap<string, StaticValidator>,
  schemaName: string,
  definition: string,
): StaticValidator {
  const validator = definitions.get(definition);
  if (validator) {
    return validator;
  }
  throw new Error(`Unknown ${schemaName} schema definition: ${definition}.`);
}

/** Returns a bounded structural summary that never echoes the validated content. */
function describeErrors(validator: StaticValidator): string {
  const errors = validator.errors ?? [];
  return errors
    .slice(0, MAXIMUM_REPORTED_ERRORS)
    .map((error) => `${error.instancePath || "/"} ${error.keyword}`)
    .join("; ");
}

/** Validates one decoded JSON value against the canonical envelope schema. */
export function isValidMessage(value: unknown): value is RinoIpcMessageV1 {
  return messageValidator(value);
}

/** Returns a bounded validation summary for diagnostics without echoing payload data. */
export function describeMessageErrors(value: unknown): string {
  if (messageValidator(value)) {
    return "";
  }
  return describeErrors(messageValidator);
}

/** Validates a message payload or result against one named canonical definition. */
export function isValidPayload(
  definition: PayloadDefinitionName,
  value: unknown,
): boolean {
  return definitionValidator(
    schemaDefinitionValidators.ipc,
    "IPC",
    definition,
  )(value);
}

/** Validates one persisted project document against the canonical graph schema. */
export function isValidProjectDocument(
  value: unknown,
): value is RinoProjectDocumentV1 {
  return projectDocumentValidator(value);
}

/** Returns a bounded validation summary for a rejected project document. */
export function describeProjectDocumentErrors(value: unknown): string {
  if (projectDocumentValidator(value)) {
    return "";
  }
  return describeErrors(projectDocumentValidator);
}

/** Validates one registry snapshot against the canonical registry schema. */
export function isValidRegistrySnapshot(
  value: unknown,
): value is RinoNodeRegistrySnapshotV1 {
  return registrySnapshotValidator(value);
}

/** Returns a bounded validation summary for a rejected registry snapshot. */
export function describeRegistrySnapshotErrors(value: unknown): string {
  if (registrySnapshotValidator(value)) {
    return "";
  }
  return describeErrors(registrySnapshotValidator);
}

/** Validates one persisted project.rino.json manifest. */
export function isValidProjectManifest(
  value: unknown,
): value is ProjectManifestV1 {
  return definitionValidator(
    schemaDefinitionValidators.graph,
    "graph",
    "ProjectManifestV1",
  )(value);
}

/** Returns a bounded validation summary for a rejected project manifest. */
export function describeProjectManifestErrors(value: unknown): string {
  const validator = definitionValidator(
    schemaDefinitionValidators.graph,
    "graph",
    "ProjectManifestV1",
  );
  return validator(value) ? "" : describeErrors(validator);
}

/** Validates one persisted graph file. */
export function isValidGraphDocument(value: unknown): value is GraphDocumentV1 {
  return definitionValidator(
    schemaDefinitionValidators.graph,
    "graph",
    "GraphDocumentV1",
  )(value);
}

/** Returns a bounded validation summary for a rejected graph file. */
export function describeGraphDocumentErrors(value: unknown): string {
  const validator = definitionValidator(
    schemaDefinitionValidators.graph,
    "graph",
    "GraphDocumentV1",
  );
  return validator(value) ? "" : describeErrors(validator);
}

/** Validates a value against one named definition of the graph schema. */
export function isValidGraphDefinition(
  definition: string,
  value: unknown,
): boolean {
  return definitionValidator(
    schemaDefinitionValidators.graph,
    "graph",
    definition,
  )(value);
}

/** Validates a value against one named definition of the registry schema. */
export function isValidRegistryDefinition(
  definition: string,
  value: unknown,
): boolean {
  return definitionValidator(
    schemaDefinitionValidators.registry,
    "registry",
    definition,
  )(value);
}

/** Validates one graph validation report against the canonical diagnostics schema. */
export function isValidDiagnosticReport(
  value: unknown,
): value is RinoGraphDiagnosticReportV1 {
  return diagnosticReportValidator(value);
}
